### DISCUSSION TITLE
How to check whether a record exist in JS hooks

### DISCUSSION THREAD

**Discussion by jenglamlow** [pocketbase/pocketbase#3702] [Q&A] [answered]
How to check whether a row exist in a collection? Currently, I try to do this I get `sql: no rows in result set`, but the string comparison failed. What is the proper method?

```
  try {
    let record = $app.dao().findFirstRecordByFilter(collectionName, filter, params);
  } catch (err) {
    console.log(err.value);
  }
```

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand what do you mean by _"but the string comparison failed"_.

All ["Fetch single record"](https://pocketbase.io/docs/js-records/#fetch-single-record) methods should throw an error if the record you are trying to retrieve doesn't exist. Using `try/catch` is fine in this case if you just want to check for existence.

**Reply by jenglamlow** [reply]
```
 try {
    let record = $app.dao().findFirstRecordByFilter(collectionName, filter, params);
  } catch (err) {
    if (err.value === 'sql: no rows in result set') {
      console.log('does not run '); // This does not run
    }
  }
```
Is the correct approach to check for existence? Is the error only thrown for when there is no record, will there be other errors throw as well (db-related error)?

**Reply by ganigeorgiev** [reply, marked as answer]
I don't recommend relying on the raw error string as it may differ between db drivers.
In your case you can just ignore the error (the only success outcome is when the record is found).

**Comment by jenglamlow** [comment]
Follow up question, how to do upsert in JS hooks?

**Reply by ganigeorgiev** [reply]
Currently there is no dedicated method for upsert queries.

In the above code-sample you can try to perform the record create inside the `catch` block.

Optionally, you can also wrap the 2 queries (select + create on select fail) in a [transaction](https://pocketbase.io/docs/js-database/#transaction) but note that in SQLite mixed read&write transactions will block the execution of other queries and I'd recommend to avoid or at least keep them at minimum to prevent performance issues.
